{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "771613cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import json\n",
    "from pathlib import Path, PurePath\n",
    "from anndata import AnnData\n",
    "from matplotlib.image import imread"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "02a5d338",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_visium(\n",
    "    path: Path | str,\n",
    "    genome: str | None = None,\n",
    "    *,\n",
    "    count_file: str = \"filtered_feature_bc_matrix.h5\",\n",
    "    library_id: str | None = None,\n",
    "    load_images: bool | None = True,\n",
    "    source_image_path: Path | str | None = None,\n",
    ") -> AnnData:\n",
    "    \"\"\"\\\n",
    "    Read 10x-Genomics-formatted visum dataset.\n",
    "\n",
    "    In addition to reading regular 10x output,\n",
    "    this looks for the `spatial` folder and loads images,\n",
    "    coordinates and scale factors.\n",
    "    Based on the `Space Ranger output docs`_.\n",
    "\n",
    "    See :func:`~scanpy.pl.spatial` for a compatible plotting function.\n",
    "\n",
    "    .. _Space Ranger output docs: https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/overview\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    path\n",
    "        Path to directory for visium datafiles.\n",
    "    genome\n",
    "        Filter expression to genes within this genome.\n",
    "    count_file\n",
    "        Which file in the passed directory to use as the count file. Typically would be one of:\n",
    "        'filtered_feature_bc_matrix.h5' or 'raw_feature_bc_matrix.h5'.\n",
    "    library_id\n",
    "        Identifier for the visium library. Can be modified when concatenating multiple adata objects.\n",
    "    source_image_path\n",
    "        Path to the high-resolution tissue image. Path will be included in\n",
    "        `.uns[\"spatial\"][library_id][\"metadata\"][\"source_image_path\"]`.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Annotated data matrix, where observations/cells are named by their\n",
    "    barcode and variables/genes by gene name. Stores the following information:\n",
    "\n",
    "    :attr:`~anndata.AnnData.X`\n",
    "        The data matrix is stored\n",
    "    :attr:`~anndata.AnnData.obs_names`\n",
    "        Cell names\n",
    "    :attr:`~anndata.AnnData.var_names`\n",
    "        Gene names for a feature barcode matrix, probe names for a probe bc matrix\n",
    "    :attr:`~anndata.AnnData.var`\\\\ `['gene_ids']`\n",
    "        Gene IDs\n",
    "    :attr:`~anndata.AnnData.var`\\\\ `['feature_types']`\n",
    "        Feature types\n",
    "    :attr:`~anndata.AnnData.obs`\\\\ `[filtered_barcodes]`\n",
    "        filtered barcodes if present in the matrix\n",
    "    :attr:`~anndata.AnnData.var`\n",
    "        Any additional metadata present in /matrix/features is read in.\n",
    "    :attr:`~anndata.AnnData.uns`\\\\ `['spatial']`\n",
    "        Dict of spaceranger output files with 'library_id' as key\n",
    "    :attr:`~anndata.AnnData.uns`\\\\ `['spatial'][library_id]['images']`\n",
    "        Dict of images (`'hires'` and `'lowres'`)\n",
    "    :attr:`~anndata.AnnData.uns`\\\\ `['spatial'][library_id]['scalefactors']`\n",
    "        Scale factors for the spots\n",
    "    :attr:`~anndata.AnnData.uns`\\\\ `['spatial'][library_id]['metadata']`\n",
    "        Files metadata: 'chemistry_description', 'software_version', 'source_image_path'\n",
    "    :attr:`~anndata.AnnData.obsm`\\\\ `['spatial']`\n",
    "        Spatial spot coordinates, usable as `basis` by :func:`~scanpy.pl.embedding`.\n",
    "    \"\"\"\n",
    "    path = Path(path)\n",
    "    adata = sc.read_10x_h5(path / count_file, genome=genome)\n",
    "\n",
    "    adata.uns[\"spatial\"] = dict()\n",
    "\n",
    "    from h5py import File\n",
    "\n",
    "    with File(path / count_file, mode=\"r\") as f:\n",
    "        attrs = dict(f.attrs)\n",
    "    if library_id is None:\n",
    "        library_id = str(attrs.pop(\"library_ids\")[0], \"utf-8\")\n",
    "\n",
    "    adata.uns[\"spatial\"][library_id] = dict()\n",
    "\n",
    "    if load_images:\n",
    "        tissue_positions_file = (\n",
    "            path / \"spatial/tissue_positions.csv\"\n",
    "            if (path / \"spatial/tissue_positions.csv\").exists()\n",
    "            else path / \"spatial/tissue_positions_list.csv\"\n",
    "        )\n",
    "        files = dict(\n",
    "            tissue_positions_file=tissue_positions_file,\n",
    "            scalefactors_json_file=path / \"spatial/scalefactors_json.json\",\n",
    "            lowres_image=path / \"spatial/tissue_lowres_image.png\",\n",
    "        )\n",
    "\n",
    "        # check if files exists, continue if images are missing\n",
    "        for f in files.values():\n",
    "            if not f.exists():\n",
    "                if any(x in str(f) for x in [\"lowres_image\"]):\n",
    "                    logg.warning(\n",
    "                        f\"You seem to be missing an image file.\\n\"\n",
    "                        f\"Could not find '{f}'.\"\n",
    "                    )\n",
    "                else:\n",
    "                    raise OSError(f\"Could not find '{f}'\")\n",
    "\n",
    "                    \n",
    "        adata.uns[\"spatial\"][library_id][\"images\"] = dict()\n",
    "        for res in [\"lowres\"]:\n",
    "            try:\n",
    "                adata.uns[\"spatial\"][library_id][\"images\"][res] = imread(\n",
    "                    str(files[f\"{res}_image\"])\n",
    "                )\n",
    "            except Exception:\n",
    "                raise OSError(f\"Could not find '{res}_image'\")\n",
    "\n",
    "        # read json scalefactors\n",
    "        adata.uns[\"spatial\"][library_id][\"scalefactors\"] = json.loads(\n",
    "            files[\"scalefactors_json_file\"].read_bytes()\n",
    "        )\n",
    "\n",
    "        adata.uns[\"spatial\"][library_id][\"metadata\"] = {\n",
    "            k: (str(attrs[k], \"utf-8\") if isinstance(attrs[k], bytes) else attrs[k])\n",
    "            for k in (\"chemistry_description\", \"software_version\")\n",
    "            if k in attrs\n",
    "        }\n",
    "\n",
    "        # read coordinates\n",
    "        positions = pd.read_csv(\n",
    "            files[\"tissue_positions_file\"],\n",
    "            header=0 if tissue_positions_file.name == \"tissue_positions.csv\" else None,\n",
    "            index_col=0,\n",
    "        )\n",
    "        positions.columns = [\n",
    "            \"in_tissue\",\n",
    "            \"array_row\",\n",
    "            \"array_col\",\n",
    "            \"pxl_col_in_fullres\",\n",
    "            \"pxl_row_in_fullres\",\n",
    "        ]\n",
    "\n",
    "        adata.obs = adata.obs.join(positions, how=\"left\")\n",
    "\n",
    "        adata.obsm[\"spatial\"] = adata.obs[\n",
    "            [\"pxl_row_in_fullres\", \"pxl_col_in_fullres\"]\n",
    "        ].to_numpy()\n",
    "        adata.obs.drop(\n",
    "            columns=[\"pxl_row_in_fullres\", \"pxl_col_in_fullres\"],\n",
    "            inplace=True,\n",
    "        )\n",
    "\n",
    "        # put image path in uns\n",
    "        if source_image_path is not None:\n",
    "            # get an absolute path\n",
    "            source_image_path = str(Path(source_image_path).resolve())\n",
    "            adata.uns[\"spatial\"][library_id][\"metadata\"][\"source_image_path\"] = str(\n",
    "                source_image_path\n",
    "            )\n",
    "\n",
    "    return adata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9122a7da",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D:\\anaconda\\envs\\index\\lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "D:\\anaconda\\envs\\index\\lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "C:\\Users\\23147\\AppData\\Local\\Temp\\ipykernel_12476\\3382230323.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  meta_temp.rename(index=lambda s: s.split('_')[-1], inplace=True)\n",
      "D:\\anaconda\\envs\\index\\lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "D:\\anaconda\\envs\\index\\lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "C:\\Users\\23147\\AppData\\Local\\Temp\\ipykernel_12476\\3382230323.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  meta_temp.rename(index=lambda s: s.split('_')[-1], inplace=True)\n",
      "D:\\anaconda\\envs\\index\\lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "D:\\anaconda\\envs\\index\\lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "C:\\Users\\23147\\AppData\\Local\\Temp\\ipykernel_12476\\3382230323.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  meta_temp.rename(index=lambda s: s.split('_')[-1], inplace=True)\n",
      "D:\\anaconda\\envs\\index\\lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "D:\\anaconda\\envs\\index\\lib\\site-packages\\anndata\\_core\\anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "C:\\Users\\23147\\AppData\\Local\\Temp\\ipykernel_12476\\3382230323.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  meta_temp.rename(index=lambda s: s.split('_')[-1], inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[AnnData object with n_obs × n_vars = 747 × 24356\n",
       "     obs: 'in_tissue', 'array_row', 'array_col', 'orig.ident', 'nCount_Spatial', 'nFeature_Spatial', 'percent.mito', 'Spatial_snn_res.1', 'seurat_clusters', 'scanorama_snn_res.1', 'Cardiomyocytes-1', 'Cardiomyocytes-2', 'Immature.myocardial.cells', 'Valve.cells', 'Macrophages', 'Fibroblast.cells', 'Erythrocytes', 'Endocardial.cells', 'MT-enriched.cardiomyocytes', 'Epi-epithelial.cells', 'Vascular.endothelial.cells', 'TMSB4X.high.cells', 'Epi-mesenchymal.cells', 'Dendritic.cells', 'Mural.cells', 'max', 'celltype_prediction_max', 'celltype_prediction', 'celltype_prediction_mode', 'region'\n",
       "     var: 'gene_ids', 'feature_types', 'genome'\n",
       "     uns: 'spatial'\n",
       "     obsm: 'spatial',\n",
       " AnnData object with n_obs × n_vars = 1966 × 24356\n",
       "     obs: 'in_tissue', 'array_row', 'array_col', 'orig.ident', 'nCount_Spatial', 'nFeature_Spatial', 'percent.mito', 'Spatial_snn_res.1', 'seurat_clusters', 'scanorama_snn_res.1', 'Cardiomyocytes-1', 'Cardiomyocytes-2', 'Immature.myocardial.cells', 'Valve.cells', 'Macrophages', 'Fibroblast.cells', 'Erythrocytes', 'Endocardial.cells', 'MT-enriched.cardiomyocytes', 'Epi-epithelial.cells', 'Vascular.endothelial.cells', 'TMSB4X.high.cells', 'Epi-mesenchymal.cells', 'Dendritic.cells', 'Mural.cells', 'max', 'celltype_prediction_max', 'celltype_prediction', 'celltype_prediction_mode', 'region'\n",
       "     var: 'gene_ids', 'feature_types', 'genome'\n",
       "     uns: 'spatial'\n",
       "     obsm: 'spatial',\n",
       " AnnData object with n_obs × n_vars = 1916 × 24356\n",
       "     obs: 'in_tissue', 'array_row', 'array_col', 'orig.ident', 'nCount_Spatial', 'nFeature_Spatial', 'percent.mito', 'Spatial_snn_res.1', 'seurat_clusters', 'scanorama_snn_res.1', 'Cardiomyocytes-1', 'Cardiomyocytes-2', 'Immature.myocardial.cells', 'Valve.cells', 'Macrophages', 'Fibroblast.cells', 'Erythrocytes', 'Endocardial.cells', 'MT-enriched.cardiomyocytes', 'Epi-epithelial.cells', 'Vascular.endothelial.cells', 'TMSB4X.high.cells', 'Epi-mesenchymal.cells', 'Dendritic.cells', 'Mural.cells', 'max', 'celltype_prediction_max', 'celltype_prediction', 'celltype_prediction_mode', 'region'\n",
       "     var: 'gene_ids', 'feature_types', 'genome'\n",
       "     uns: 'spatial'\n",
       "     obsm: 'spatial',\n",
       " AnnData object with n_obs × n_vars = 1967 × 24356\n",
       "     obs: 'in_tissue', 'array_row', 'array_col', 'orig.ident', 'nCount_Spatial', 'nFeature_Spatial', 'percent.mito', 'Spatial_snn_res.1', 'seurat_clusters', 'scanorama_snn_res.1', 'Cardiomyocytes-1', 'Cardiomyocytes-2', 'Immature.myocardial.cells', 'Valve.cells', 'Macrophages', 'Fibroblast.cells', 'Erythrocytes', 'Endocardial.cells', 'MT-enriched.cardiomyocytes', 'Epi-epithelial.cells', 'Vascular.endothelial.cells', 'TMSB4X.high.cells', 'Epi-mesenchymal.cells', 'Dendritic.cells', 'Mural.cells', 'max', 'celltype_prediction_max', 'celltype_prediction', 'celltype_prediction_mode', 'region'\n",
       "     var: 'gene_ids', 'feature_types', 'genome'\n",
       "     uns: 'spatial'\n",
       "     obsm: 'spatial']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_dir = \"G:/dataset/13-ChickenHeart/input/\"\n",
    "samples = ['D4', 'D7', 'D10', 'D14']\n",
    "meta = pd.read_csv(r'G:\\dataset\\13-ChickenHeart\\input\\spatialRNAseq_metadata.csv', index_col=0)\n",
    "\n",
    "adata_list = []\n",
    "for sample in samples:\n",
    "    adata = read_visium(input_dir + sample)\n",
    "    meta_temp = meta[meta['orig.ident']==sample]\n",
    "    meta_temp.rename(index=lambda s: s.split('_')[-1], inplace=True)\n",
    "    adata.obs = pd.concat([adata.obs,meta_temp],axis=1)\n",
    "    adata_list.append(adata)\n",
    "    \n",
    "adata_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2d28a925",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(adata_list)):\n",
    "    adata = adata_list[i]\n",
    "    cells=adata.obs_names\n",
    "    mat=pd.DataFrame(adata.X.todense(),index=cells,columns=adata.var_names)\n",
    "    coord=pd.DataFrame(adata.obsm['spatial'],index=cells,columns=['x','y'])\n",
    "    meta=adata.obs[['region','orig.ident']]\n",
    "    meta.columns=['celltype','batch']\n",
    "    meta.index=cells\n",
    "    mat.to_csv(input_dir+samples[i]+\"_mat.csv\")\n",
    "    meta.to_csv(input_dir+samples[i]+\"_meta.csv\")\n",
    "    coord.to_csv(input_dir+samples[i]+\"_coord.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "06a25078",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(4):\n",
    "    adata_list[i].write_h5ad(input_dir + samples[i] + '.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "index",
   "language": "python",
   "name": "index"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
